### ISSUE TITLE
Optimized shape inference with AdvancedSubtensor1 + Reshape

### ISSUE THREAD

**Issue by sisp** [Theano/Theano#2567]
I noticed that shape inference optimization does not work properly when I perform advanced indexing with a higher-dimensional index variable. Currently, `AdvancedSubtensor1` is used in this case along with two `Reshape` ops. The combination of these ops seems to prevent efficient shape inference. Consider the following code

```
import theano
import theano.tensor as T

x = T.matrix('x')
i = T.matrix('i', dtype='int64')
f = theano.function([x, i], x[i].shape[0])

theano.printing.pydotprint(f, 'graph.png', format='png')
```

which gives the following graph:
![theano](https://cloud.githubusercontent.com/assets/2206639/6526014/221c8fd4-c40d-11e4-9640-7571d5b1cb24.png)
The shape of `x[i]` can be computed without actually performing the indexing operation. I've not looked into shape optimization yet, so is it just a missing optimization here that prevents the simplification?

**Comment by lamblin** [comment]
My guess is that there is a case where the `infer_shape` method of an Op is not implemented, or is not able to determine the shape of its outputs, and simply returns `output.shape`.

**Comment by sisp** [comment]
Okay, yes, I believe the problem is in the `infer_shape` method of `Reshape`. It's exactly doing what you said unless the new shape is constant.

**Comment by sisp** [comment]
I've looked at the problem more closely and came up with a possible solution (see the diff [here](https://gist.github.com/sisp/52b9a757e76080771a5f)). Before making a PR, I'd like to discuss whether that's a viable fix.

I introduced a new op `FixUnknownDimension` which checks the sanity of the `newshape` vector and infers the missing dimension if present. Basically, this is what NumPy does. I borrowed the code of their `_fix_unknown_dimension` C function because I couldn't find a way to access it from Theano.
